File: LeetCode-Python/MaximumSumCircularSubarray.py
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        total_sum = sum(nums)
    
        # Kadane: max subarray sum (normal)
        max_sum = cur_max = nums[0]
        for num in nums[1:]:
            cur_max = max(num, cur_max + num)
            max_sum = max(max_sum, cur_max)
        
        # Kadane: min subarray sum (for circular case)
        min_sum = cur_min = nums[0]
        for num in nums[1:]:
            cur_min = min(num, cur_min + num)
            min_sum = min(min_sum, cur_min)
        
        # Eğer tüm elemanlar negatifse, circular toplam işe yaramaz
        if max_sum < 0:
            return max_sum
        
        return max(max_sum, total_sum - min_sum)

File: LeetCode-Cpp/MaximumSumCircularSubarray.cpp
class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        int totalSum = 0;
        int maxSum = nums[0], curMax = nums[0];
        int minSum = nums[0], curMin = nums[0];

        for (int i = 1; i < nums.size(); ++i) {
            int num = nums[i];
            // Kadane's algorithm for max subarray sum
            curMax = max(num, curMax + num);
            maxSum = max(maxSum, curMax);

            // Kadane's algorithm for min subarray sum
            curMin = min(num, curMin + num);
            minSum = min(minSum, curMin);

            totalSum += num;
        }

        totalSum += nums[0]; // nums[0] was skipped in the loop

        // Eğer tüm elemanlar negatifse, circular toplam işe yaramaz
        if (maxSum < 0) return maxSum;

        // max of normal max subarray or circular subarray
        return max(maxSum, totalSum - minSum);

    }
};
